cpp/src/arrow/compute/kernels/aggregate_test.cc

```diff
@@ -189,6 +189,53 @@ TEST(TestBooleanAggregation, Sum) {
               ResultWith(Datum(MakeNullScalar(uint64()))));
 }
 
+TEST(TestBooleanAggregation, Product) {
+  const ScalarAggregateOptions& options = ScalarAggregateOptions::Defaults();
+  ValidateBooleanAgg<Product>("[]", std::make_shared<UInt64Scalar>(), options);
```

Comment:
FindAccumulatorType has a specialization for Boolean so that it's UInt64Scalar: https://github.com/apache/arrow/blob/2aa94c4712ce406d7c87d361b5c655a6ea585701/cpp/src/arrow/compute/kernels/aggregate_internal.h#L32-L35

Reply:
Ok, but kernel registration does the following:
```c++
  aggregate::AddArrayScalarAggKernels(aggregate::ProductInit::Init, {boolean()}, int64(),
                                      func.get());
```

(same for `Sum`, incidentally)

Reply:
Well, that needs fixing :grimacing: good catch!